modules/xobjdetect/include/opencv2/xobjdetect/gp_regression.hpp

```diff
@@ -0,0 +1,102 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+/******************************************************************************
+ *
+ * This software is the core implementation of the fine-grained search (FGS) method
+ * based on the Gaussian process regression with latent variables.
+ *
+ * Details of the method can be found in the follow paper
+ *   Yuting Zhang, Kihyuk Sohn, Ruben Villegas, Gang Pan, Honglak Lee
+ *   ``Improving Object Detection with Deep Convolutional Networks via Bayesian Optimization and Structured Prediction''
+ *   IEEE Conference on Computer Vision and Pattern Recognition (CVPR), Boston, MA, 2015.
+ *
+ * The following people contributed to the code:
+ *   Yuting Zhang   <zyt.cse@gmail.com>
+ *   Kihyuk Sohn    <kihyuks@umich.edu>
+ *   Ruben Villegas <rubville@umich.edu>
+ *   Xinchen Yan    <xcyan@umich.edu>
+ *   Gang Pan       <gpan@zju.edu.cn>
+ *   Honglak Lee    <honglak@eecs.umich.edu>
+ *
+*******************************************************************************/
+
+#ifndef INCLUDE_OPENCV2_FGS_GP_REGRESSION_HPP_
+#define INCLUDE_OPENCV2_FGS_GP_REGRESSION_HPP_
+
+#include <string>
+#include <cstdio>
+#include <iostream>
+#include <fstream>
+
+#include "opencv2/core.hpp"
+
+namespace cv2 {
+
+// argmax z on joint probability of the existing boxes
+// argmax yNp1 on conditional probability
+
+//    kNp1: N x 1; KN: N x N; psiNp1: 4 x 1
+
+class fgs_gp_box_reg {
+protected:
+    /**
+     * variables
+     */
+     struct GPmodel_dfn {
+        double m0;
+        cv::Mat diagSqrtLambda; // 4 x 1
+        double normCov;
+        double noiseSigma2;
+        cv::Mat idxbScaleEnabled; // 4 x 1
+    } GPmodel_;
+    //cv2::fgs_base_detector& detector_;
+
+    /**
+     * private/protected functions
+     */
+    cv::Mat sgp_kNp1_forward(const cv::Mat &, const cv::Mat &, const GPmodel_dfn &);
```

Comment:
all the internal methods and class members should be put to the implementation class